console.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "CONSOLE" "1" "March 2013" "" ""
.
.SH "NAME"
\fBconsole\fR \- A Read\-Eval\-Print Loop (REPL) for the PHP language\.
.
.SH "SYNOPSIS"
\fBconsole\fR [\fB\-hHqz\fR] [\fB\-c\fR \fIcolor\fR] [\fB\-f\fR \fIfile\fR] [\fB\-i\fR \fIkey=val\fR] [\fB\-p\fR \fIfile\fR] [\fB\-s\fR \fIfile\fR] [\fB\-v\fR \fIvar=val\fR]
.
.SH "DESCRIPTION"
\fBConsole\fR provides a Read\-Eval\-Print Loop (REPL) environment for the PHP language\. A REPL is an interactive programming environment\. The user enters PHP statements and expressions which are read, evaluated, and the results printed\. The REPL environment is protected from errors and exceptions, including PHP fatal and parse errors\.
.
.SH "OPTIONS"
.
.TP
\fB\-A\fR
Don\'t preload tab\-completion cache\-\-\-console starts up a little faster but tab completion is a little slower\.
.
.TP
\fB\-c\fR \fIcolor\fR
Return value print color (default is \fBcyan\fR)\. Choices are: \fBblack\fR, \fBred\fR, \fBgreen\fR, \fByellow\fR, \fBblue\fR, \fBmagenta\fR, \fBcyan\fR, \fBwhite\fR, and \fBnone\fR\. Bold colors can be specified as \fBbold\-\fR\fIcolor\fR\. Specify \fBnone\fR to disable colored output\.
.
.TP
\fB\-f\fR \fIfile\fR
Require \fIfile\fR before starting REPL\.
.
.TP
\fB\-h\fR
Print usage info and exit\.
.
.TP
\fB\-H\fR
Don\'t parse \fB\.htaccess\fR files at startup\.
.
.TP
\fB\-i\fR \fIkey=val\fR
Set PHP configuration option \fIkey\fR to \fIval\fR\.
.
.TP
\fB\-p\fR \fIfile\fR
Require \fIfile\fR before loading console\'s classloader\.
.
.TP
\fB\-q\fR
Don\'t echo the result after evaling each expression\.
.
.TP
\fB\-s\fR \fIfile\fR
Run console commands in \fIfile\fR before interactive REPL\.
.
.TP
\fB\-v\fR \fIvar=val\fR
Set \fI$var\fR to \fIval\fR globally in REPL environment\.
.
.TP
\fB\-z\fR
Run script files but don\'t start interactive REPL\.
.
.P
Multiple \fB\-f\fR, \fB\-i\fR, \fB\-p\fR, and \fB\-v\fR options may be specified on the same command line\.
.
.SH "REPL COMMANDS"
The following commands are available inside the REPL environment:
.
.TP
\fB/d\fR \fIthing\fR
Get the doc comment for the \fIthing\fR\.
.
.TP
\fB/e\fR [\fIfile\fR]
Append session history to \fIfile\fR and open in editor\. If \fIfile\fR is not specified then the file specified with the \fB\-s\fR option will be used\.
.
.TP
\fB/f\fR \fIfile\fR
Load \fIfile\fR via php\'s \fBrequire()\fR\.
.
.TP
\fB/h\fR
Print this help info\.
.
.TP
\fB/hh\fR
Print expression history\.
.
.TP
\fB/p\fR \fIexpr\fR
Toggle echoing the result just for this expression\.
.
.TP
\fB/pp\fR
Toggle echoing the result of each eval\.
.
.TP
\fB/pager\fR [\fIcmd\fR]
Pipe results through shell command \fIcmd\fR when displaying them\. If no \fIcmd\fR is provided, the current pager command is removed\.
.
.TP
\fB/q\fR \fIexpr\fR
Evaluate \fIexpr\fR but don\'t display the result\.
.
.TP
\fB/x\fR \fIexpr\fR
Evaluate \fIexpr\fR and examine result (print full \fBvar_export\fR output)\.
.
.SH "HISTORY EXPANSION"
Each statement and expression entered into the REPL is assigned a history number\. Expressions and statements previously entered can be referenced by this number at read\-time via history expansion\. The following special forms are expanded:
.
.TP
\fB$$\fR
Expands to the previous statement/expression\.
.
.TP
\fB$\fR\fIN\fR
Expands to the statement/expression corresponding to history number \fIN\fR\.
.
.TP
\fB$\fR\fI\-N\fR
Expands to the \fIN\fRth previous statement/expression\. (Note that \fB$$\fR is equivalent to \fB$\-1\fR, for example\.)
.
.TP
\fB$!\fR
Expands to the previous result\. (The previous statement or expression is not re\-evaluated\.)
.
.SH "CONFIGURATION"
Most of the REPL functionality can be modified or accessed programmatically via the \fBConsole\fR and \fBConsoleCommand\fR classes\.
.
.TP
\fBConsole::$DEBUG\fR
Extra debugging info is printed when this static property is \fBtrue\fR\.
.
.TP
\fBConsole::$PAGER\fR
Shell command to pipe results through (e\.g\. "tee /tmp/foo | more")\.
.
.TP
\fBConsole::$PS1\fR
This public static property controls how the console prompt is presented\. It may contain either a \fBprintf(\fR3\fB)\fR style format string or a callback function\. The format string or function should accept 3 arguments: history number, elapsed time, and memory used\. The callback function should return the string to be used as the prompt\.
.
.TP
\fBConsole::$PS2\fR
This public static property is a string to be used for the secondary (continued line) prompt\.
.
.TP
\fBConsole::$printers\fR
This is an array whose keys are class names and whose values are functions\. When the console displays an object of the given class, the function will be called with one argument\-\-\-the class name\. The function should return the string to be printed\. If a printer is not found for the object\'s class, then the printer for its parent class is used, if one exists, and so on\.
.
.TP
\fBConsoleCommand::add(\fRstring \fI$name\fR, callable \fI$f\fR\fB)\fR
Add a command to the REPL, accessible via \fB/\fR\fI$name\fR\. The \fI$f\fR callback should accept one argument, the rest of the command line, and it should return a string (which will be evaluated) or nothing\.
.
.TP
\fBConsoleCommand::doit(\fRstring \fI$line\fR\fB)\fR
Programmatically run the REPL command \fI$line\fR\.
.
.SH "FILES"
Console looks for the php file \fB~/\.consolerc\.php\fR\. If it finds it then that file is loaded via \fBrequire_once\fR when the REPL is started\. Additionally, the REPL can be configured via \fB\.htaccess\fR files, as usual with the Apache web server\.
.
.P
The console requires a scratch file \fB~/\.console\.php\.scratch\fR for the pager if a pager is specified, and a history file \fB~/\.console\.php\.history\fR if expression history is used\.
.
.SH "COPYRIGHT"
Wigwam is Copyright (c) 2012 Micha Niskin
.
.P
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.
.P
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\.
.
.P
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE\.


Console/Console.php
<?php namespace Wigwam\Console;

use \RuntimeException;
use \ReflectionMethod;

function substr_replace_first($str, $find, $replace) {
  return (($pos = strpos($str, $find)) !== false)
    ? substr_replace($str, $replace, $pos, strlen($find))
    : $str;
}

function num_tok($arr) {
  return is_array($arr) && $arr[0] == T_LNUMBER ? $arr[1] : null;
}

/**
 * This is the Console class.
 */
class Console {
 
  /* ANSI terminal color escape */
  public static $colors = array(
    "none"          => -1,
    "black"         => '0;30',
    "red"           => '0;31',
    "green"         => '0;32',
    "yellow"        => '0;33',
    "blue"          => '0;34',
    "magenta"       => '0;35',
    "cyan"          => '0;36',
    "white"         => '0;37',
    "bold-black"    => '1;30',
    "bold-red"      => '1;31',
    "bold-green"    => '1;32',
    "bold-yellow"   => '1;33',
    "bold-blue"     => '1;34',
    "bold-magenta"  => '1;35',
    "bold-cyan"     => '1;36',
    "bold-white"    => '1;37',
  );

  public static $PS1_COLOR    = "\033[34m[\033[33m%s\033[34m] [\033[33m%.3fs\033[34m] [\033[33m%.1fMB\033[34m]\033[0m\n>>> ";

  /** Current prompt. */
  public static $n            = 0;
  public static $PS1          = "[%s] [%.3fs] [%.1fMB]\n>>> ";
  public static $PS2          = '+++ ';
  public static $OUTCOLOR     = '1;36';
  public static $welcome;

  public static $INTERACTIVE  = true;
  public static $DIE_HANDLER;
  public static $EXC_HANDLER;

  public static $HISTORY_RES;
  public static $HISTSIZE     = 1000;

  public static $HISTFILE_S   = "";
  public static $SCRIPTFILE   = "";
  public static $RUNSCRIPT    = array();

  public static $SH;
  public static $PAGER;

  public static $DEBUG        = false;

  public static $getopt       = "f:q";
  public static $option       = array();

  public static $print          = true;
  public static $printnext      = true;
  public static $printshort     = true;
  public static $printshortnext = true;

  public static $sock         = array();
  public static $reboot       = false;
  public static $die          = false;
  public static $pid1;
  public static $pid2;
  public static $pid3;
  public static $status;
  
  public static $line;
  public static $result;
  public static $no_result;

  public static $tmp;
  public static $tmp_result;
  public static $tmp_status   = array(0,0);
  public static $start_time   = 0.0;
  public static $last_time    = 0.0;
  public static $mem_usage    = 0.0;

  public static $completion_error;
  public static $ignore_errors              = false;
  public static $print_before_error_message = "";
  public static $print_after_error_message  = "";

  public static $printers = array();

  public static $argv;

  private static $printable_tokens = array(
    '!',
    '+',
    '-',
    '(',
    '~',
    '@',
    '`',
    T_ARRAY,
    T_ARRAY_CAST,
    T_BOOL_CAST,
    T_CLASS_C,
    T_CLONE,
    T_CONSTANT_ENCAPSED_STRING,
    T_DEC,
    T_DIR,
    T_DNUMBER,
    T_DOUBLE_CAST,
    T_EMPTY,
    T_ENCAPSED_AND_WHITESPACE,
    T_FILE,
    T_FUNC_C,
    T_INC,
    T_INT_CAST,
    T_ISSET,
    T_LINE,
    T_LIST,
    T_LNUMBER,
    T_METHOD_C,
    T_NS_C,
    T_NEW,
    T_NUM_STRING,
    T_OBJECT_CAST,
    T_PRINT,
    T_STRING,
    T_STRING_CAST,
    T_STRING_VARNAME,
    T_UNSET_CAST,
    T_VARIABLE,
  );

  public static function emitTryCatch() {
    while(array_key_exists(
      ($name = str_replace('.', '', uniqid('____wigwam_',true))), $GLOBALS));

    $expr = <<<EOT
      try {
        Wigwam\\Console\\Console::color("bold-white");
        Wigwam\\Console\\Console::\$tmp_result = eval(Wigwam\\Console\\Console::getLine());
      } catch (\\Exception \$__NAME__) {
        Wigwam\\Console\\Console::printResult(\$__NAME__);
        Wigwam\\Console\\Console::\$result    = \$__NAME__;
        Wigwam\\Console\\Console::printException(\$__NAME__);
        Wigwam\\Console\\Console::\$printnext = false;
      }

      unset(\$__NAME__);
EOT;

    return str_replace('__NAME__', $name, $expr);
  }

  public static function probe() {
    $args = func_get_args();
    $cl   = array_shift($args);
    $meth = array_shift($args);

    $r = new ReflectionMethod($cl, $meth);
    $r->setAccessible(true);

    return call_user_func_array(
      array($r, "invoke"),
      array_merge(array(is_object($cl) ? $cl : null), $args));
  }

  public static function done() {
    Console::$ignore_errors = true;
    die();
  }

  public static function prompt() {
    $p = static::$PS1;
    $n = static::$n;
    $t = static::$last_time;
    $m = static::$mem_usage;
    return is_callable($p) ? $p($n, $t, $m) : sprintf($p, $n, $t, $m);
  }

  public static function strcolor($color="none") {
    $color = $color == "none" ? 0 : static::$colors[$color];
    return static::$OUTCOLOR != -1 ? "\033[{$color}m" : "";
  }

  public static function color($color="none") {
    if (static::$OUTCOLOR != -1)
      printf(static::strcolor($color));
  }

  public static function center($s) {
    $w    = exec('tput cols');
    $s    = array_map(function($x) use ($w) {
              return substr($x, 0, $w);
            }, array_map('rtrim', explode("\n", $s)));
    $max  = array_reduce(array_map('strlen', $s), 'max', 0);
    $pad  = $w > $max ? str_repeat(' ', floor(($w - $max) / 2) - 1) : '';

    return $pad . implode("\n$pad", $s) . "\n";
  }

  public static function errMsg($message, $file, $line) {
    return sprintf(
      "%sPHP [%d]: %s\nIn %s line %d%s",
      Console::$print_before_error_message,
      posix_getpid(),
      $message,
      $file,
      $line,
      Console::$print_after_error_message
    );
  }

  public static function printException($e) {
    $t = $e->getTrace();
    $p = $e->getPrevious();

    Console::color("bold-red");

    if ($e->getFile() != $t[0]['file'] || $e->getLine() != $t[0]['line'])
      error_log(sprintf("AT %s(%d)", $e->getFile(), $e->getLine()));
    error_log(implode("\n", array_slice(explode("\n", $e->getTraceAsString()), 0, -4)));

    Console::color();

    if ($p) {
      Console::color("yellow");
      error_log(Console::pp($p));
      Console::color();
      Console::printException($p);
    }
  }

  public static function printErr($message, $file, $line) {
    Console::color("bold-red");
    error_log(static::errMsg($message, $file, $line));
    Console::color();
  }

  public static function printFatal($e) {
    static::printErr($e['message'], $e['file'], $e['line']);
  }

  public static function setup() {
    global $argv;

    error_reporting(0);

    static::$SH = trim(`which sh`);
    static::getSocks();

    static::$argv = $argv;

    if (! count(readline_list_history())) {
      $h = unserialize(file_get_contents(static::getHistFile()));
      $h = is_array($h) ? $h : array();

      readline_clear_history();

      array_map(
        'readline_add_history',
        array_reverse(array_slice(array_reverse($h), 0, static::$HISTSIZE)));
    }

    static::$n = count(readline_list_history());

    register_shutdown_function(function() {
      if (! Console::$ignore_errors && $e = error_get_last()) {
        if (Console::$DIE_HANDLER) {
          $h = Console::$DIE_HANDLER;
          $e = array_merge(
            array('type'=>null,'message'=>null,'file'=>null,'line',null),
            ifnil(error_get_last(), array()));
          $h($e['type'], $e['message'], $e['file'], $e['line']);
        } else
          Console::printFatal($e);
      }
      die();
    });

    set_error_handler(function ($errno, $errstr, $errfile, $errline) {
      if (Console::$DIE_HANDLER) {
        $h = Console::$DIE_HANDLER;
        $h($errno, $errstr, $errfile, $errline);
      } else
        Console::printErr($errstr, $errfile, $errline);
      Console::done();
    });

    set_exception_handler(function($e) {
      if (Console::$EXC_HANDLER) {
        $h = Console::$EXC_HANDLER;
        $h($e);
      } else
        Console::printException($e);
      Console::done();
    });

    $completion_fn = function($buf, $i) use (&$completion_fn) {
      $orig   = $buf;
      $info   = readline_info();
      $point  = $info['point'];
      $buf    = substr($info['line_buffer'], 0, $point);

      if ($point && $buf[$point-1] == ' ') {
        readline_info('pending_input', chr(8));
        readline_callback_read_char();
        return;
      }

      // Prevent empty completion request from being sent, as it would cause
      // the socket to block forever waiting for the response.
      if (! $buf) return;

      $match  = Console::getCompletionResp(Console::sendCompletionReq($buf));

      if (is_array($match) && count($match)) {
        sort($match);
        return $match;
      }
    };
    readline_completion_function($completion_fn);
  }

  public static function fork(&$pid) {
    if (($pid = pcntl_fork()) < 0)
      throw new RuntimeException("can't fork");
  }

  public static function getSocks() {
    if (! socket_create_pair(AF_UNIX, SOCK_STREAM, 0, static::$sock[0]))
      throw new RuntimeException("can't create socket pair");
    if (! socket_create_pair(AF_UNIX, SOCK_STREAM, 0, static::$sock[1]))
      throw new RuntimeException("can't create socket pair");
  }

  public static function readSock($i, $j) {
    $buflen = 8192;
    $sock   = static::$sock[$i][$j];
    $bin    = PHP_BINARY_READ;

    for($ret='';;) {
      if (($buf = socket_read($sock, $buflen, $bin)) === false)
        throw new RuntimeException("can't read from socket");
      $ret.=$buf;
      if ($buf == '' || strlen($buf) < $buflen)
        break;
    }

    return $ret;
  }

  public static function writeSock($i, $j, $buf) {
    if (socket_write(static::$sock[$i][$j], $buf) === false)
      throw new RuntimeException("can't write to socket");
  }

  public static function getHistFile() {
    $path = $_SERVER['HOME']."/.console.php.history";
    $path = is_writable(dirname($path)) ? $path : "/dev/null";
    touch($path);
    return $path;
  }

  public static function gobbleWhitespace(&$toks) {
    while (count($toks) && count($toks[0]) && $toks[0][0] == T_WHITESPACE)
      array_shift($toks);
  }

  public static function expandHistory($line) {
    $toks = token_get_all('<?php '.$line);
    if (! count($toks))
      return;

    $h = readline_list_history();

    array_shift($toks);

    for ($i=0; $i<count($toks)-1; $i++) {
      $t1 = $toks[$i];
      $t2 = $toks[$i+1];

      if ($toks[$i] == '$') {

        // Substitute history result in expression.
        if (! is_null($n = num_tok($toks[$i+1])))
          return static::expandHistory(
            substr_replace_first(
              $line, "\${$n}", $h[$n]));
        if ($toks[$i+1] == '$')
          return static::expandHistory(
            substr_replace_first(
              $line, "\$\$", $h[count($h)-1]));
        if ($toks[$i+1] == '!')
          return static::expandHistory(
            substr_replace_first(
              $line, "\$!", "Wigwam\\Console\\Console::\$result"));
        if ($i < count($toks)-2 && $toks[$i+1] == '-' &&
          ! is_null($n = num_tok($toks[$i+2])))
          return static::expandHistory(
            substr_replace_first(
              $line, "\$-{$n}", $h[count($h)-$n]));
      }
    }

    return $line;
  }

  public static function printableLine($line) {
    $toks = token_get_all('<?php '.$line);
    if (! count($toks))
      return;

    array_shift($toks);

    static::gobbleWhitespace($toks);

    if (static::$DEBUG)
      error_log(var_export($toks, true));

    if (count($toks) && count($toks[0]) && $toks[0][0] == T_FUNCTION) {
      array_shift($toks);
      static::gobbleWhitespace($toks);
      return $toks[0] == '(';
    }

    return 
      (count($toks) &&
        (count($toks[0]) &&
          (in_array($toks[0][0], static::$printable_tokens) ||
          in_array($toks[0], static::$printable_tokens))))
        ? $line
        : false;
  }

  public static function escapeHist($arg) {
    return addcslashes(addcslashes(escapeshellarg($arg), "\r\n\t"), '\\');
  }

  public static function isValidPHP($line) {
    $f = tempnam(sys_get_temp_dir(),"");
    $c = "php -l '$f'";
    file_put_contents($f, "<?php\n$line\n?>\n");
    exec($c, $out);
    unlink($f);
    return count($out) && substr($out[0],0,2) == 'No';
  }

  public static function doReadline($prompt) {
    $lines  = explode("\n", $prompt);
    $prompt = array_pop($lines);
    echo(implode("\n", $lines)."\n");

    $line = static::expandHistory(readline($prompt));

    if ($line && ($l = substr(ltrim($line),0,1)) != '/') {
      while (strlen($l) && ! static::isValidPHP($line)) {
        $l = static::expandHistory(readline(static::$PS2));
        $line .= "\n$l";
      }

      if (static::isValidPHP($line)) {
        readline_add_history($line);
        static::writeHistory();
        static::readHistory();
      }
    } elseif ($line === false) {
      static::$reboot = true;
      $line = '/q Wigwam\\Console\\Console::$reboot = true';
    }

    return strlen($line) ? $line : ' ';
  }

  public static function stopWorkers() {
    static::writeSock(0, 0, ":done:");
    static::writeSock(1, 0, ":done:");
  }

  public static function readHistory() {
    $h = unserialize(file_get_contents(static::getHistFile()));
    $h = is_array($h) ? $h : array();
    readline_clear_history();
    array_map('readline_add_history', $h);
    Console::$n = count(readline_list_history());
  }

  public static function writeHistory() {
    file_put_contents(
      static::getHistFile(),
      serialize(array_reverse(array_slice(
        array_reverse(readline_list_history()), 0, static::$HISTSIZE))));
  }

  public static function readLine() {
    if (count(static::$RUNSCRIPT)) {
      $line = array_shift(static::$RUNSCRIPT);
      if (Console::$INTERACTIVE)
        echo static::prompt()."$line";
    } else {
      if (! Console::$INTERACTIVE) {
        $line = '/q Wigwam\\Console\\Console::$die = Wigwam\\Console\\Console::$reboot = true';
        static::$reboot = true;
        static::$die    = true;
      } else
        $line = static::doReadline(static::prompt());
    }
    static::writeSock(0, 0, $line);
  }

  public static function getLine() {
    $line = preg_replace('/;$/', '', static::readSock(0, 1));

    if (static::$completion_error)
      $line = " ";

    if (strlen(trim($line)) && static::isValidPHP($line)) {
      static::readHistory();
    }

    $line = ConsoleCommand::doit($line);
    $plin = static::printableLine($line);

    $line = 'Wigwam\Console\Console::$result = '
      . ($plin ? '' : 'Wigwam\Console\Console::$no_result; ')
      . $line . ';';

    static::$line = $line;

    return $line;
  }

  public static function printObj($o) {
    $r = new \ReflectionClass($o);
    $p = Console::$printers;

    while ($r && !array_key_exists($r->getName(), $p))
      $r = $r->getParentClass();

    $pp = $r ? $p[$r->getName()] : function($o) {
      return sprintf("#<%s>", get_class($o));
    };

    return $pp($o);
  }

  public static function pp($thing, $shift='  ') {
    $ret = "";
    if (is_array($thing) && ! count($thing)) {
      $ret .= "array()";
    } elseif (is_array($thing)) {
      $ret .= "array(\n";
      foreach ($thing as $k => $v)
        $ret .= sprintf(
          "%s%s => %s,\n",
          $shift,
          var_export($k, true),
          implode("\n$shift", explode("\n", static::pp($v, $shift)))
        );
      $ret .= ")";
    } elseif (is_object($thing)) {
      $ret .= Console::printObj($thing);
    } else {
      $ret .= var_export($thing, true);
    }

    return $ret;
  }

  public static function outColor() {
    if (static::$INTERACTIVE && static::$OUTCOLOR != -1)
      printf("\033[%sm", static::$OUTCOLOR);
  }

  public static function printResult($res) {
    if (static::$printnext) {
      $res = static::$printshortnext
        ? static::pp($res)
        : var_export($res, true);

      static::outColor();

      if (static::$INTERACTIVE && static::$PAGER && is_writable($_SERVER['HOME'])) {
        $scratch = $_SERVER['HOME']."/.console.php.scratch";
        file_put_contents($scratch, $res);
        $pid = pcntl_fork();
        if ($pid > 0) {
          pcntl_wait($status);
          return;
        } elseif (! $pid) {
          $cmd = sprintf("cat %s 2>&- |%s", escapeshellarg($scratch), static::$PAGER);
          pcntl_exec(static::$SH, array("-c", $cmd));
        } else {
          $ex = new \Exception("Can't fork.");
          printf(static::pp($ex)."\n");
          throw $ex;
        }
      } else
        printf("%s\n", $res);

      static::color();
    }
    static::$printshortnext = static::$printshort;
    static::$printnext = static::$print;
  }

  public static function getStatus() {
    return static::readSock(0, 0);
  }

  public static function sendStatus($status) {
    static::writeSock(0, 1, $status);
  }

  public static function sendCompletionReq($buf) {
    static::writeSock(1, 0, $buf);
  }

  public static function getCompletionReq() {
    return static::readSock(1, 1);
  }

  public static function sendCompletionResp($match) {
    static::writeSock(1, 1, serialize($match));
  }

  public static function getCompletionResp() {
    return unserialize(static::readSock(1, 0));
  }

  public static function serviceCompletionRequests() {
    static::$completion_error = false;

    if (! socket_create_pair(AF_UNIX, SOCK_STREAM, 0, $sock))
      throw new RuntimeException("can't create socket pair");

    socket_set_nonblock($sock[1]);

    while (($tmp = static::getCompletionReq()) != ':done:') {
      $pid = pcntl_fork();

      if (! $pid) {
        // child

        Console::$print_before_error_message = "\n";

        $forceStatic = preg_match('/^\\/d /', $tmp);
        $c = new ConsoleCommandCompletion($forceStatic);
        static::sendCompletionResp($c->complete($tmp));

        if (socket_write($sock[0], "ok") === false)
          throw new RuntimeException("can't write to socket");
        
        Console::done();
      } else {
        // parent

        pcntl_waitpid($pid, $status, WUNTRACED);

        $buf = socket_read($sock[1], 8192, PHP_BINARY_READ);

        if ($buf != "ok") {
          static::$completion_error = true;
          //static::endCompletions();
          static::sendCompletionResp("ERROR");
        }
      }
    }

  }

  public static function endCompletions() {
    static::sendCompletionReq(':done:');
  }
}
